### PR TITLE
New sparse image subresource sparse storage for sync, implementing image_layout map

### PR DESCRIPTION
**Pull request by jzulauf-lunarg**
Implements an range map and subresource adapters suitable for synchronization work, integrated as a replacement for the ImageSubresourceLayout map to test for robustness, correctness, and performance. Eliminates uses of ForRange in favor of a combine iterator for image layout                                                   validation.

With initial hot-spot profiling/optimization and performance savings w.r.t. eliminating new/free needed for `ForRange` std::function and the lambda captures, initial testing indicates rough performance parity with the previous sparse_vector approach for image_layout tracking and is designed to scale to "all device memory" byte resolution tracking for sync.

The google style is used for the image_layout and subresource_adapter namespaces.  The sparse container is intended to be in an STL style.

### CODE REVIEW

**Comment by jzulauf-lunarg** [comment]
@jeffbolznv  -- if you have any performance test case you'd like to run against this, that would be great.  I think it should be have performance parity with the prior approach overall, but as I'm planning to use the subresource_adapter and range_vector piece for synchronization, it would be nice to find any glass-jaws early.

File: layers/buffer_validation.cpp [hunk 1]
```diff
@@ -3436,9 +3427,9 @@ bool CoreChecks::ValidateCmdBufImageLayouts(const CMD_BUFFER_STATE *pCB, const I
         isr_pair.image = image;
         isr_pair.hasSubresource = true;
         // Validate the initial_uses for each subresource referenced
-        for (auto it_init = subres_map->BeginInitialUse(); !it_init.AtEnd(); ++it_init) {
-            isr_pair.subresource = (*it_init).subresource;
-            VkImageLayout initial_layout = (*it_init).layout;
+        for (const auto &pos : subres_map->InitialLayoutView()) {
+            isr_pair.subresource = pos.subresource;
+            VkImageLayout initial_layout = pos.it->second;  // Point directly into the RangeMap iterated but the view
```
Comment (mark-lunarg) [inline layers/buffer_validation.cpp]: Does this mean the rangemap except for the view?

File: layers/image_layout_map.cpp [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+            if (!parallel_it->range.empty() && !parallel_it->range.includes(current)) {
+                ++parallel_it;
+            }
+        }
+        // ++range_gen will update subres_gen.
+    }
+    return keep_on;
+}
+
+// This is the same constant value range, subreource position advance logic as ForRange above, but suitable for use with
```
Comment (mark-lunarg) [inline layers/image_layout_map.cpp]: subreource

File: layers/image_layout_map.h [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+        VkImageLayout initial_layout;
+        bool operator==(const SubresourceRangeLayout& rhs) const {
+            bool is_equal = (current_layout == rhs.current_layout) && (initial_layout == rhs.initial_layout) &&
+                            (subresource_range == rhs.subresource_range);
+            return is_equal;
+        }
+        bool operator!=(const SubresourceRangeLayout& rhs) const { return !(*this == rhs); }
+    };
+
+    using RangeMap = sparse_container::range_map<IndexType, VkImageLayout>;
```
Comment (mark-lunarg) [inline layers/image_layout_map.h]: grumblegrumblegrumble

File: layers/CMakeLists.txt [hunk 4]
```diff
@@ -162,7 +162,12 @@ set(CHASSIS_LIBRARY_FILES
     generated/chassis.cpp
     generated/layer_chassis_dispatch.cpp
     generated/command_counter_helper.cpp
-    state_tracker.cpp)
+    state_tracker.cpp
```
Comment (mark-lunarg) [inline layers/CMakeLists.txt]: I think I understand why these are here --  it'd be more clear if the statetracker stuff was in a state_tracker_files group.

**Comment by jzulauf-lunarg** [comment]
@jeffbolznv  -- the perf work took longer than I planned... if you've got a change to re-review re-test.  Also, if your test procedure is something you can share, please do.

@locke-lunarg  -- the image layout code here overall profile about the same speed as the code it replaces (overall), if you can check against your baseline testing, that would be useful to make sure my test case wasn't atypical.

**Comment by jzulauf-lunarg** [comment]
@locke-lunarg  -- branch updated with fixes from CTS test run.

**Comment by jzulauf-lunarg** [comment]
@jeffbolznv  -- did you ever get a chance to look at the perf tuned version of this.  It clears CTS and I was going to rebase and merge, then add @locke-lunarg change that converts the global map handling of layout tracking to the new container (with more local perf/CTS checking) as a separate PR).

**Comment by jeffbolznv** [comment]
Sorry, I missed the update. I'll try to test it soon.

**Pull request merged**